tawazi/_dag/helpers.py

```diff
@@ -0,0 +1,222 @@
+from concurrent.futures import ALL_COMPLETED, FIRST_COMPLETED, Future, ThreadPoolExecutor, wait
+from copy import copy
+from typing import Any, Dict, List, Optional, Set
+
+from loguru import logger
+
+from tawazi._dag.digraph import DiGraphEx
+from tawazi.consts import Identifier
+from tawazi.errors import ErrorStrategy
+from tawazi.node.node import ExecNode
+
+
+def _xn_active_in_call(xn: ExecNode, xns_dict: Dict[Identifier, ExecNode]) -> bool:
+    if isinstance(xn.active, bool):
+        return xn.active
+    return bool(xns_dict[xn.active.id].result)
+
+
+def copy_non_setup_xns(x_nodes: Dict[str, ExecNode]) -> Dict[str, ExecNode]:
+    """Deep copy all ExecNodes except setup ExecNodes because they are shared throughout the DAG instance.
+
+    Args:
+        x_nodes: Dict[str, ExecNode] x_nodes to be deep copied
+
+    Returns:
+        Dict[str, ExecNode] copy of x_nodes
+    """
+    x_nodes_copy = {}
+    for id_, x_nd in x_nodes.items():
+        # if execnode is a setup node, it shouldn't be copied
+        if x_nd.setup:
+            x_nodes_copy[id_] = x_nd
+        else:
+            # no need to deepcopy. we only need to know if self.result is NoVal or not (TODO: fix this COmment)
+            x_nodes_copy[id_] = copy(x_nd)
+    return x_nodes_copy
+
+
+# 0.5 create helpers functions encapsulated from the outside
```

Comment:
is the "0.5" still necessary in this case ?

Reply:
Fixed